### PR TITLE
Adds validation for workflow task for child start and cancel

### LINKED ISSUE #276: Fix handling of immediate child workflow cancellation
Currently when a single decision batch contains `StartChildWorkflowExecution` and `RequestCancelExternalWorkflowExecution` the cancellation fails and the child is started.
The expected behaviour is that the child is immediately cancelled ideally by writing the following events into the history:

1. `RequestCancelExternalWorkflowExecutionInitiated`
2. `ExternalWorkflowExecutionCancelRequested`
3. `ChildWorkflowExecutionCanceled`

Here is an example history which shows the current broken behaviour:
```
   1  2020-04-05T16:34:53-07:00  EventTypeWorkflowExecutionStarted                         {WorkflowType:{Name:TestWorkflow_execute}, ParentInitiatedEventId:0,
                                                                                           TaskList:{Name:WorkflowTest-testChildWorkflowCancellation-49a80f10-fb1d-49c6-a9f1-282b2e31f460,
                                                                                           Kind:TaskListKindNormal}, Input:["WAIT_CANCELLATION_REQUESTED"],
                                                                                           ExecutionStartToCloseTimeoutSeconds:30, TaskStartToCloseTimeoutSeconds:5,
                                                                                           Initiator:ContinueAsNewInitiatorDecider, ContinuedFailureDetails:[], LastCompletionResult:[],
                                                                                           OriginalExecutionRunId:f2bc6299-a8b3-4996-9c0c-f4414f1a79bd, Identity:unknown-mac,
                                                                                           FirstExecutionRunId:f2bc6299-a8b3-4996-9c0c-f4414f1a79bd, Attempt:0, ExpirationTimestamp:0,
                                                                                           FirstDecisionTaskBackoffSeconds:0}
   2  2020-04-05T16:34:53-07:00  EventTypeDecisionTaskScheduled                            {TaskList:{Name:WorkflowTest-testChildWorkflowCancellation-49a80f10-fb1d-49c6-a9f1-282b2e31f460,
                                                                                           Kind:TaskListKindNormal}, StartToCloseTimeoutSeconds:5, Attempt:0}
   3  2020-04-05T16:34:53-07:00  EventTypeWorkflowExecutionCancelRequested                 {ExternalInitiatedEventId:0,
                                                                                           Identity:unknown-mac}
   4  2020-04-05T16:34:53-07:00  EventTypeDecisionTaskStarted                              {ScheduledEventId:2, Identity:unknown-mac,
                                                                                           RequestId:15e68a7d-672f-4e20-937c-3f3b3c646734}
   5  2020-04-05T16:34:53-07:00  EventTypeDecisionTaskCompleted                            {ExecutionContext:[],
                                                                                           ScheduledEventId:2,
                                                                                           StartedEventId:4,
                                                                                           Identity:unknown-mac}
   6  2020-04-05T16:34:53-07:00  EventTypeStartChildWorkflowExecutionInitiated             {WorkflowId:68563d49-3289-3c7a-9a39-e0e0d21d39bf,
                                                                                           WorkflowType:{Name:TestChildWorkflow_execute},
                                                                                           TaskList:{Name:WorkflowTest-testChildWorkflowCancellation-49a80f10-fb1d-49c6-a9f1-282b2e31f460,
                                                                                           Kind:TaskListKindNormal}, Input:[], ExecutionStartToCloseTimeoutSeconds:30,
                                                                                           TaskStartToCloseTimeoutSeconds:10, ParentClosePolicy:ParentClosePolicyAbandon,
                                                                                           Control:[], DecisionTaskCompletedEventId:5,
                                                                                           WorkflowIdReusePolicy:WorkflowIdReusePolicyAllowDuplicateFailedOnly}
   7  2020-04-05T16:34:53-07:00  EventTypeRequestCancelExternalWorkflowExecutionInitiated  {DecisionTaskCompletedEventId:5,
                                                                                           WorkflowExecution:{WorkflowId:68563d49-3289-3c7a-9a39-e0e0d21d39bf},
                                                                                           Control:[], ChildWorkflowOnly:true}
   8  2020-04-05T16:34:53-07:00  EventTypeRequestCancelExternalWorkflowExecutionFailed     {Cause:CancelExternalWorkflowExecutionFailedCauseUnknownExternalWorkflowExecution,
                                                                                           DecisionTaskCompletedEventId:-23, Namespace:UnitTest,
                                                                                           WorkflowExecution:{WorkflowId:68563d49-3289-3c7a-9a39-e0e0d21d39bf}, InitiatedEventId:7,
                                                                                           Control:[]}
   9  2020-04-05T16:34:53-07:00  EventTypeDecisionTaskScheduled                            {TaskList:{Name:unknown-mac:83e37a46-43b3-40e0-bbc0-a40f27208f84,
                                                                                           Kind:TaskListKindNormal}, StartToCloseTimeoutSeconds:5,
                                                                                           Attempt:0}
  10  2020-04-05T16:34:53-07:00  EventTypeChildWorkflowExecutionStarted                    {InitiatedEventId:6,
                                                                                           WorkflowExecution:{WorkflowId:68563d49-3289-3c7a-9a39-e0e0d21d39bf,
                                                                                           RunId:c4b71487-52e5-4984-a47d-105fa4478979},
                                                                                           WorkflowType:{Name:TestChildWorkflow_execute}}
  11  2020-04-05T16:34:53-07:00  EventTypeDecisionTaskStarted                              {ScheduledEventId:9,
                                                                                           Identity:83e37a46-43b3-40e0-bbc0-a40f27208f84,
                                                                                           RequestId:995457b0-b937-4333-b5c1-2b6083bee769}
  12  2020-04-05T16:34:53-07:00  EventTypeDecisionTaskCompleted                            {ExecutionContext:[],
                                                                                           ScheduledEventId:9,
                                                                                           StartedEventId:11,
                                                                                           Identity:unknown-mac}
  13  2020-04-05T16:35:23-07:00  EventTypeWorkflowExecutionTimedOut                        {TimeoutType:TimeoutTypeStartToClose}
```

### PR DESCRIPTION
**Pull request by samarabbas**
**What changed?**

Adds validation to fail workflow task when child execution is started
and cancelled in the same workflow task.

Fixes #276 

**Why?**

If client SDK starts a child workflow and then cancels it in the same workflow task Temporal
currently processes both these commands which results in the child workflow getting started
and then cancellation requested on it.  This results in undesired behavior on the sdk, where much
better expectation is that child workflow is never started in the first place.

Initially we thought of implementing logic on the server to cancel the child execution immediately as
part of the workflow task processing.  But after more thought it is not as simple as expected.  Most
of the code around child workflows state machine is completely asynchronous and short circuiting 
all that logic to synchronous cancel the child execution could be very risky.

So after discussion with @mfateev I decided to consider this as an invalid scenario and fail the entire 
workflow task instead.  We would add logic on the clients to never make this command if child is
started and cancelled in the same workflow task.

**How did you test it?**
Added new integration test `TestImmediateChildCancellation_WorkflowTaskFailed`

**Potential risks**
This state machine already results in broken experience on SDK.  Does not introduces any additional risk.

### CODE REVIEW

File: service/history/workflowTaskHandler.go [hunk 1]
```diff
@@ -96,12 +97,13 @@ func newWorkflowTaskHandler(
...
-		continueAsNewBuilder:        nil,
-		stopProcessing:              false,
-		mutableState:                mutableState,
+		hasBufferedEvents:                 mutableState.HasBufferedEvents(),
+		failWorkflowTaskInfo:              nil,
+		activityNotStartedCancelled:       false,
+		continueAsNewBuilder:              nil,
+		stopProcessing:                    false,
+		mutableState:                      mutableState,
+		initiatedChildExecutionsInSession: make(map[string]struct{}),
```
Comment (mfateev) [inline service/history/workflowTaskHandler.go:106]: Do we have a similar problem with other commands? What about activities, timers, etc?

File: service/history/workflowTaskHandler.go [hunk 2]
```diff
@@ -54,12 +54,13 @@ type (
...
-		continueAsNewBuilder        mutableState
-		stopProcessing              bool // should stop processing any more commands
-		mutableState                mutableState
+		hasBufferedEvents                 bool
+		failWorkflowTaskInfo              *failWorkflowTaskInfo
+		activityNotStartedCancelled       bool
+		continueAsNewBuilder              mutableState
+		stopProcessing                    bool // should stop processing any more commands
+		mutableState                      mutableState
+		initiatedChildExecutionsInSession map[string]struct{} // Set of initiated child executions in the workflow task
```
Comment (mfateev) [inline service/history/workflowTaskHandler.go:63]: session is a new term. Should we call more like currentCommandBatch or something else?

Comment (samarabbas) [reply service/history/workflowTaskHandler.go:63]: Changed to `initiatedChildExecutionsInBatch`

File: service/history/workflowTaskHandler.go (hunk 1 above)
Comment (samarabbas) [reply service/history/workflowTaskHandler.go]: I will add integration tests for timer and activity use cases in a separate PR.

File: host/cancelworkflow_test.go [hunk 3]
```diff
@@ -534,3 +536,172 @@ CheckHistoryLoopForCancelSent:
...
+			WorkflowId: childWorkflowID,
+		},
+	})
+	if err == nil {
+		s.printWorkflowHistory(s.namespace, &commonpb.WorkflowExecution{
+			WorkflowId: childWorkflowID,
+		})
+	}
+	s.Logger.Error("Describe error", tag.Error(err))
+	s.NotNil(err, "Child workflow execution started instead of getting cancelled")
```
Comment (alexshtin) [inline host/cancelworkflow_test.go:703]: nit:
```suggestion
	s.Error(err, "Child workflow execution started instead of getting cancelled")
```

**Pull request merged**